File: scripts/ai_council/project_context.py
"""Lightweight project-context detector for the council handoff preamble.

Council members do better critique when they know what the project IS,
not just what the artefact looks like. This module reads the bare
minimum from the repo root — `composer.json`, `package.json`, root
`README.md` — and returns a neutral `ProjectContext`. All fields are
optional; missing data is `None` and the preamble silently omits the
line.

Iron law of neutrality (`ai-council` skill): nothing here may carry
host-agent identity, prior reasoning, or framing. Manifest fields and
README prose only.
"""

from __future__ import annotations

import json
import re
from dataclasses import dataclass
from pathlib import Path

REPO_PURPOSE_MAX_CHARS = 400
_HEADING_RE = re.compile(r"^\s*#")
_BADGE_RE = re.compile(r"^\s*(\[!\[|!\[|<).*")
_HTML_RE = re.compile(r"<[^>]+>")


@dataclass
class ProjectContext:
    """Neutral project description for the council handoff preamble."""

    name: str | None = None
    stack: str | None = None
    repo_purpose: str | None = None

    def is_empty(self) -> bool:
        return self.name is None and self.stack is None and self.repo_purpose is None


def _read_json(path: Path) -> dict | None:
    if not path.exists():
        return None
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except (json.JSONDecodeError, OSError):
        return None
    return data if isinstance(data, dict) else None


def _name_from(composer: dict | None, package: dict | None, root: Path) -> str | None:
    for src in (composer, package):
        if src and isinstance(src.get("name"), str) and src["name"].strip():
            return src["name"].strip()
    # Fall back to the directory name; useful for repos without manifests.
    try:
        return root.resolve().name or None
    except OSError:
        return None


def _stack_from(composer: dict | None, package: dict | None) -> str | None:
    parts: list[str] = []
    if composer:
        php_v = (composer.get("require") or {}).get("php")
        if isinstance(php_v, str):
            parts.append(f"PHP {php_v}")
        # Detect well-known frameworks without claiming the project IS one.
        require = {**(composer.get("require") or {}), **(composer.get("require-dev") or {})}
        for needle, label in (
            ("laravel/framework", "Laravel"),
            ("symfony/framework-bundle", "Symfony"),
            ("laminas/laminas-mvc", "Laminas"),
        ):
            if needle in require:
                parts.append(label)
                break
    if package:
        engines = package.get("engines") or {}
        if isinstance(engines, dict) and isinstance(engines.get("node"), str):
            parts.append(f"Node {engines['node']}")
        deps = {**(package.get("dependencies") or {}), **(package.get("devDependencies") or {})}
        for needle, label in (
            ("next", "Next.js"),
            ("react", "React"),
            ("vue", "Vue"),
            ("@angular/core", "Angular"),
        ):
            if needle in deps:
                parts.append(label)
                break
    if not parts:
        return None
    return " · ".join(parts)


def _purpose_from_readme(path: Path) -> str | None:
    if not path.exists():
        return None
    try:
        text = path.read_text(encoding="utf-8")
    except OSError:
        return None
    paragraph: list[str] = []
    for raw in text.splitlines():
        line = raw.rstrip()
        stripped = line.strip()
        if not stripped:
            if paragraph:
                break
            continue
        if _HEADING_RE.match(stripped) or _BADGE_RE.match(stripped):
            if paragraph:
                break
            continue
        paragraph.append(stripped)
    if not paragraph:
        return None
    joined = " ".join(paragraph)
    joined = _HTML_RE.sub("", joined).strip()
    if not joined:
        return None
    if len(joined) > REPO_PURPOSE_MAX_CHARS:
        joined = joined[: REPO_PURPOSE_MAX_CHARS - 1].rstrip() + "…"
    return joined


def detect_project_context(root: Path | None = None) -> ProjectContext:
    """Return a `ProjectContext` for `root` (default: cwd).

    Always returns — never raises. Missing manifest files / README → the
    matching field is `None`, and `handoff_preamble()` will omit the line.
    """
    root = (root or Path.cwd()).resolve()
    composer = _read_json(root / "composer.json")
    package = _read_json(root / "package.json")
    return ProjectContext(
        name=_name_from(composer, package, root),
        stack=_stack_from(composer, package),
        repo_purpose=_purpose_from_readme(root / "README.md"),
    )

File: src/scripts/ai_council/project_context.ts
/**
 * Lightweight project-context detector for the council handoff preamble.
 *
 * TypeScript twin of `src/scripts/ai_council/project_context.py` (ADR-200 —
 * Python→TS migration, Phase 1). Reads the bare minimum from the repo root —
 * `composer.json`, `package.json`, root `README.md` — and returns a neutral
 * `ProjectContext`. All fields are optional; missing data is `null` and the
 * preamble silently omits the line.
 *
 * Iron law of neutrality (`ai-council` skill): nothing here may carry
 * host-agent identity, prior reasoning, or framing. Manifest fields and
 * README prose only.
 *
 * Truncation strategy (locked by council review, 2026-05-02): `repo_purpose`
 * is capped at `REPO_PURPOSE_MAX_CHARS` by stopping at the last full sentence
 * ≤ 400 chars, with an ellipsis when truncation occurred. Never cut
 * mid-sentence.
 */
import * as fs from 'node:fs';
import * as path from 'node:path';

export const REPO_PURPOSE_MAX_CHARS = 400;

// Python: re.compile(r"^\s*#"), re.compile(r"^\s*(\[!\[|!\[|<).*"),
// re.compile(r"<[^>]+>"). re.match anchors at the start (not ^$ MULTILINE);
// these are applied to single stripped lines so anchoring is implicit.
const _HEADING_RE = /^\s*#/;
const _BADGE_RE = /^\s*(\[!\[|!\[|<).*/;
const _HTML_RE = /<[^>]+>/g; // re.sub replaces all → global flag.

/** Neutral project description for the council handoff preamble. */
export class ProjectContext {
    name: string | null;
    stack: string | null;
    repo_purpose: string | null;

    constructor(
        name: string | null = null,
        stack: string | null = null,
        repoPurpose: string | null = null,
    ) {
        this.name = name;
        this.stack = stack;
        this.repo_purpose = repoPurpose;
    }

    is_empty(): boolean {
        return this.name === null && this.stack === null && this.repo_purpose === null;
    }
}

type JsonDict = Record<string, unknown>;

function _isPlainObject(v: unknown): v is JsonDict {
    return typeof v === 'object' && v !== null && !Array.isArray(v);
}

function _read_json(p: string): JsonDict | null {
    // Python: if not path.exists(): return None
    if (!fs.existsSync(p)) {
        return null;
    }
    let raw: string;
    try {
        raw = fs.readFileSync(p, 'utf-8');
    } catch {
        // OSError → None
        return null;
    }
    let data: unknown;
    try {
        data = JSON.parse(raw);
    } catch {
        // json.JSONDecodeError → None
        return null;
    }
    // Python: return data if isinstance(data, dict) else None
    return _isPlainObject(data) ? data : null;
}

function _name_from(
    composer: JsonDict | null,
    pkg: JsonDict | null,
    root: string,
): string | null {
    for (const src of [composer, pkg]) {
        if (src) {
            const nameVal = src['name'];
            if (typeof nameVal === 'string' && nameVal.trim() !== '') {
                return nameVal.trim();
            }
        }
    }
    // Fall back to the directory name; useful for repos without manifests.
    // Python: root.resolve().name or None  (root is already resolved by caller).
    try {
        const base = path.basename(root);
        return base || null;
    } catch {
        // OSError → None
        return null;
    }
}

function _asObject(v: unknown): JsonDict {
    // Python `(x.get("y") or {})` — falsy (None, missing) → {}. Here: only a
    // plain object survives; anything else (null, undefined, non-dict) → {}.
    return _isPlainObject(v) ? v : {};
}

function _stack_from(composer: JsonDict | null, pkg: JsonDict | null): string | null {
    const parts: string[] = [];
    if (composer) {
        const require0 = _asObject(composer['require']);
        const phpV = require0['php'];
        if (typeof phpV === 'string') {
            parts.push(`PHP ${phpV}`);
        }
        // Detect well-known frameworks without claiming the project IS one.
        const requireMerged: JsonDict = {
            ..._asObject(composer['require']),
            ..._asObject(composer['require-dev']),
        };
        for (const [needle, label] of [
            ['laravel/framework', 'Laravel'],
            ['symfony/framework-bundle', 'Symfony'],
            ['laminas/laminas-mvc', 'Laminas'],
        ] as Array<[string, string]>) {
            if (needle in requireMerged) {
                parts.push(label);
                break;
            }
        }
    }
    if (pkg) {
        const engines = _asObject(pkg['engines']);
        // Python: isinstance(engines, dict) and isinstance(engines.get("node"), str)
        if (typeof engines['node'] === 'string') {
            parts.push(`Node ${engines['node'] as string}`);
        }
        const deps: JsonDict = {
            ..._asObject(pkg['dependencies']),
            ..._asObject(pkg['devDependencies']),
        };
        for (const [needle, label] of [
            ['next', 'Next.js'],
            ['react', 'React'],
            ['vue', 'Vue'],
            ['@angular/core', 'Angular'],
        ] as Array<[string, string]>) {
            if (needle in deps) {
                parts.push(label);
                break;
            }
        }
    }
    if (parts.length === 0) {
        return null;
    }
    return parts.join(' · ');
}

/** Mirror Python `str.splitlines()` — splits on the universal newline set. */
function _splitlines(text: string): string[] {
    if (text === '') {
        return [];
    }
    // Python splitlines() boundaries include LF CR CRLF VT FF FS GS RS
    // NEL LS PS. README content in practice only carries LF / CRLF, but
    // replicate the full set for faithfulness.
    const parts = text.split(/\r\n|[\n\r\v\f\x1c\x1d\x1e\x85\u2028\u2029]/u);
    if (parts.length > 0 && parts[parts.length - 1] === '') {
        parts.pop();
    }
    return parts;
}

function _purpose_from_readme(p: string): string | null {
    if (!fs.existsSync(p)) {
        return null;
    }
    let text: string;
    try {
        text = fs.readFileSync(p, 'utf-8');
    } catch {
        // OSError → None
        return null;
    }
    const paragraph: string[] = [];
    for (const raw of _splitlines(text)) {
        // Python: line = raw.rstrip(); stripped = line.strip()
        const stripped = raw.trim();
        if (!stripped) {
            if (paragraph.length > 0) {
                break;
            }
            continue;
        }
        if (_HEADING_RE.test(stripped) || _BADGE_RE.test(stripped)) {
            if (paragraph.length > 0) {
                break;
            }
            continue;
        }
        paragraph.push(stripped);
    }
    if (paragraph.length === 0) {
        return null;
    }
    let joined = paragraph.join(' ');
    joined = joined.replace(_HTML_RE, '').trim();
    if (!joined) {
        return null;
    }
    if (_pyLen(joined) > REPO_PURPOSE_MAX_CHARS) {
        joined = _truncate_at_sentence(joined, REPO_PURPOSE_MAX_CHARS);
    }
    return joined;
}

/** Mirror Python `len(str)` — code-point count. */
function _pyLen(s: string): number {
    let n = 0;
    for (const _ of s) {
        n += 1;
    }
    return n;
}

/** Mirror Python `str.rstrip()` with no args — strip trailing whitespace. */
function _pyRstrip(s: string): string {
    return s.replace(/\s+$/u, '');
}

/**
 * Truncate at the last full sentence ≤ limit chars; append an ellipsis.
 * Total return length is always ≤ `limit` (ellipsis included).
 */
function _truncate_at_sentence(text: string, limit: number): string {
    const budget = Math.max(1, limit - 2);
    // Python head = text[:budget] slices by code point. JS string indexing is
    // UTF-16; slice by code points to match for astral chars.
    const head = _pySlice(text, budget);
    const cut = Math.max(_rfind(head, '. '), _rfind(head, '! '), _rfind(head, '? '));
    if (cut >= 0) {
        // Python head[: cut + 1] — cut is a code-point index from rfind below,
        // applied to the same code-point head; slice by code point again.
        return _pyRstrip(_pySliceRange(head, 0, cut + 1)) + ' …';
    }
    return _pyRstrip(head) + ' …';
}

/** Python text[:n] — first n code points. */
function _pySlice(text: string, n: number): string {
    const cps = Array.from(text);
    return cps.slice(0, n).join('');
}

/** Python text[a:b] — code-point slice [a, b). */
function _pySliceRange(text: string, a: number, b: number): string {
    const cps = Array.from(text);
    return cps.slice(a, b).join('');
}

/**
 * Python str.rfind(sub) — highest code-point index where sub starts, or -1.
 * Operates on code-point indices to match Python's slicing semantics.
 */
function _rfind(haystack: string, needle: string): number {
    const cps = Array.from(haystack);
    const hLen = cps.length;
    const nCps = Array.from(needle);
    const nLen = nCps.length;
    if (nLen === 0) {
        return hLen;
    }
    for (let i = hLen - nLen; i >= 0; i -= 1) {
        let match = true;
        for (let j = 0; j < nLen; j += 1) {
            if (cps[i + j] !== nCps[j]) {
                match = false;
                break;
            }
        }
        if (match) {
            return i;
        }
    }
    return -1;
}

/**
 * Return a `ProjectContext` for `root` (default: cwd).
 *
 * Always returns — never raises. Missing manifest files / README → the
 * matching field is `null`, and `handoff_preamble()` will omit the line.
 */
export function detect_project_context(root?: string | null): ProjectContext {
    // Python: root = (root or Path.cwd()).resolve()
    const resolvedRoot = path.resolve(root ?? process.cwd());
    const composer = _read_json(path.join(resolvedRoot, 'composer.json'));
    const pkg = _read_json(path.join(resolvedRoot, 'package.json'));
    return new ProjectContext(
        _name_from(composer, pkg, resolvedRoot),
        _stack_from(composer, pkg),
        _purpose_from_readme(path.join(resolvedRoot, 'README.md')),
    );
}
